File: packages/recorder/ios/RecorderTypes.swift
import Foundation

internal struct RecorderError: Error {
  let code: String
  let message: String
  let recoverable: Bool

  init(code: String, message: String, recoverable: Bool = false) {
    self.code = code
    self.message = message
    self.recoverable = recoverable
  }
}

internal struct RecorderConfig {
  let sampleRate: Double
  let channels: UInt32
  let bitDepth: UInt32
  let outputDirectory: String?
  let filenamePrefix: String?
  let rotateAfterMs: Double?
  let emitAudioLevel: Bool
  let audioLevelIntervalMs: Double
  let aacBitrate: Int
}

internal struct RecorderOutput {
  let uri: String
  let durationMs: Double
  let sizeBytes: Double
  let sampleRate: Double
  let channels: UInt32
  let bitDepth: UInt32
}

internal protocol RecorderCallbacks: AnyObject {
  func onState(_ state: String)
  func onAudioLevel(rms: Double, peak: Double)
  func onChunk(uri: String, startMs: Double, endMs: Double, sizeBytes: Double)
  func onError(_ error: RecorderError)
}

File: packages/recorder/android/src/main/java/com/clarionhq/recorder/RecorderTypes.kt
package com.clarionhq.recorder

internal class RecorderError(
  val code: String,
  message: String,
  val recoverable: Boolean = false,
  cause: Throwable? = null,
) : Exception(message, cause)

internal interface RecorderCallbacks {
  fun onState(state: String)
  fun onAudioLevel(rms: Double, peak: Double)
  fun onChunk(uri: String, startMs: Long, endMs: Long, sizeBytes: Long)
  fun onError(error: RecorderError)
}

internal inline fun RecorderCallbacks.runCatchingRecorder(
  fallbackCode: String,
  fallbackPrefix: String,
  block: () -> Unit,
) {
  try {
    block()
  } catch (e: RecorderError) {
    onError(e)
  } catch (e: Throwable) {
    onError(RecorderError(fallbackCode, "$fallbackPrefix: ${e.message}", cause = e))
  }
}
